### PR TITLE
Modifications to HasProperties class defaults

### PR DESCRIPTION
**Pull request by fwkoch**
- `defaults` wrapper has been removed. To implement dynamic defaults, specify a callable in the `_defaults` dictionary - this can no longer depend on `self`.
- `_defaults` dictionary is now inherited. Any defaults not specified in a class will be inherited from parent classes.
    - This differs from old behaviour where `_defaults` overwrote parent class values for `_defaults`
- default values are set on `__init__` or new function `reset`. Deleting a property sets it to undefined.
    - This differs from old behavior where defaults were set on first `get` and deleting a property set it to default.
- On init, defaults do not fire change notifications, kwargs do.
- Additional validation of handlers and defaults in the metaclass.

### CODE REVIEW

File: properties/base.py [hunk 1]
```diff
@@ -81,12 +83,20 @@ def __new__(mcs, name, bases, classdict):                                  #pyli
             prop.name = key
             classdict[key] = prop.get_property()
 
-        # Overwrite handlers with their function
+        # Ensure observed names are valid
+        for key, hand in iteritems(observer_dict):
```
Comment (ThomasMatern) [inline properties/base.py]: `prop` short for `property` - cool. `hand` short for `handler` - meh..

File: properties/base.py [hunk 2]
```diff
@@ -126,6 +138,28 @@ def __new__(mcs, name, bases, classdict):                                  #pyli
...
+                if callable(value):
+                    newcls._props[key].validate(None, value())
+                else:
+                    newcls._props[key].validate(None, value)
+            except ValueError:
+                raise AttributeError(
+                    "Invalid default for property '{}'".format(key)
+                )
+        newcls._defaults = _defaults
+
```
Comment (bsmithyman) [inline properties/base.py]: Makes sense. I wrote [something similar](https://github.com/bsmithyman/galoshes/blob/master/galoshes/meta.py#L40) in the past. Can confirm `reversed` is the right way to do this.

File: properties/base.py [hunk 3]
```diff
@@ -135,51 +169,42 @@ def __new__(mcs, name, bases, classdict):                                  #pyli
...
+        self.reset(silent=True)
+
+        # Set the keyword arguments with change notifications
+        for key, val in iteritems(kwargs):
             if not hasattr(self, key) and key not in self._props.keys():
-                raise AttributeError("Keyword input '{:s}' is not a known "
+                raise AttributeError("Keyword input '{}' is not a known "
                                      "property or attribute".format(key))
-            setattr(self, key, kwargs[key])
+            setattr(self, key, val)
```
Comment (bsmithyman) [inline properties/base.py]: @fwkoch: You were talking about maybe moving this stuff to `__new__`. Another option would be to move it to the metaclass `__call__`, as in [galoshes](https://github.com/bsmithyman/galoshes/blob/master/galoshes/meta.py#L60). That leaves the `__init__.py` undefined, and free to do whatever it wants, within a fully-initialized class. Thoughts?

File: properties/base.py [hunk 4]
```diff
@@ -190,13 +215,42 @@ def _notify(self, change):
     def _set(self, name, value):
         change = dict(name=name, value=value, mode='validate')
         self._notify(change)
-        if change['value'] is utils.undefined and name in self._backend:
-            self._backend.pop(name)
+        if change['value'] is utils.undefined:
+            self._backend.pop(name, None)
```
Comment (bsmithyman) [inline properties/base.py]: I like `.pop`. Is it for sure available in all backends? May want to add a set of "need to implement these to write a backend" notes. https://github.com/3ptscience/properties/issues/88

**Comment by bsmithyman** [comment]
@fwkoch: Maybe we want to do the `__init__` --> `mcs.__call__` switch? See comments.

File: properties/base.py [hunk 5]
```diff
@@ -135,51 +169,42 @@ def __new__(mcs, name, bases, classdict):                                  #pyli
...
-                )
-            if callable(value):
-                setattr(self, key, value())
-            else:
-                setattr(self, key, value)
-
-        # set the keywords
-        for key in kwargs:
+        # Set the GettableProperties from defaults - these are only set here
+        for key in self._props:
```
Comment (ThomasMatern) [inline properties/base.py]: You use `self._props[key]` 3 times in this loop, maybe it's a better idea to just use `iteritems` here.

File: properties/base.py (hunk 1 above)
Comment (fwkoch) [inline properties/base.py]: Hah, @bsmithyman gave me a bit of a hard time about this too. I'll use `handler`. Thanks.

File: properties/base.py (hunk 5 above)
Comment (fwkoch) [inline properties/base.py]: Good call.

File: properties/base.py (hunk 3 above)
Comment (fwkoch) [inline properties/base.py]: Implemented and approved by @rowanc1 - Thanks again @bsmithyman, this is so good.

**Pull request merged**